<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/Polars2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment: Alternatives to Pandas for Processing Large Datasets**

## **Instruction:**

You are required to compare Pandas with the selected library

Make sure you use the same dataset when making comparisons.

You can also use visualization to show the comparison.

Your work must not be plagiarized i.e. copy-pasted from somewhere else.

**Team Members:**

1.   MUHAMMAD DINIE HAZIM BIN AZALI
2.   RADIN DAFINA BINTI RADIN ZULKAR NAIN
3.   ADRINA ASYIQIN BINTI MD ADHA
4.   KELVIN EE









In [ ]:
from google.colab import drive
drive.mount('/content/drive')

**Installing Polars**

In [ ]:
pip install -U polars

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


**Importing Polars and Pandas**

In [ ]:
import polars as pl
import pandas as pd
import datetime

**Read data using pandas**

In [ ]:
%%time
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/AssignmentBigData2/Rate.csv', low_memory=False)

CPU times: user 25 s, sys: 11.2 s, total: 36.3 s
Wall time: 41.4 s


%% time is a magic action in Jupyter notebook that times all executions within the cell.

Function to get size of any object

In [ ]:
from sys import getsizeof
def whats_my_footprint(x):
    if 'x' in locals() or 'x' in globals():
        size = round(getsizeof(data) / 1024 / 1024,2)
        print(f'{size} MB in the memory')
    else:
        print('are you sure the object exists?')

pandas dataframe takes up 9517.71 MB in memory

In [ ]:
whats_my_footprint(data)

9517.71 MB in the memory


In [ ]:
%%time
display(data.RatingAreaId.value_counts())

Rating Area 1     978906
Rating Area 2     934953
Rating Area 3     920486
Rating Area 4     907736
Rating Area 5     823124
                   ...  
Rating Area 54     17309
Rating Area 62     17033
Rating Area 63     17033
Rating Area 65     17033
Rating Area 67     17033
Name: RatingAreaId, Length: 67, dtype: int64

CPU times: user 644 ms, sys: 11.5 ms, total: 655 ms
Wall time: 780 ms


In [ ]:
#Change the data type of the store flag to categorical
data.RatingAreaId = data.RatingAreaId.astype('category')

In [ ]:
%%time
display(data.dtypes)
display(whats_my_footprint(data))

BusinessYear                                    int64
StateCode                                      object
IssuerId                                        int64
SourceName                                     object
VersionNum                                      int64
ImportDate                                     object
IssuerId2                                       int64
FederalTIN                                     object
RateEffectiveDate                              object
RateExpirationDate                             object
PlanId                                         object
RatingAreaId                                 category
Tobacco                                        object
Age                                            object
IndividualRate                                float64
IndividualTobaccoRate                         float64
Couple                                        float64
PrimarySubscriberAndOneDependent              float64
PrimarySubscriberAndTwoDepen

8677.0 MB in the memory


None

CPU times: user 11 s, sys: 59.2 ms, total: 11 s
Wall time: 11 s


So, we ended up taking another few seconds to convert  categorical feature that Pandas failed to read correctly, converting it into the categorical form, and reducing footprint from 9518 MB to 8677 MB. 

In [ ]:
%%time
display(data.SourceName.value_counts())

display(whats_my_footprint(data))

HIOS     8702452
SERFF    3853855
OPM       138138
Name: SourceName, dtype: int64

8677.0 MB in the memory


None

CPU times: user 11.4 s, sys: 51.5 ms, total: 11.4 s
Wall time: 11.4 s


Next, let's see how Polar does it

**Read data using Polars**

In [ ]:
df = pl.read_csv('/content/drive/MyDrive/Colab Notebooks/AssignmentBigData2/Rate.csv')
display(df.head())

shape: (5, 24)
┌────────────┬─────────┬──────────┬──────────┬─────┬────────────┬────────────┬────────────┬─────────┐
│ BusinessYe ┆ StateCo ┆ IssuerId ┆ SourceNa ┆ ... ┆ CoupleAndO ┆ CoupleAndT ┆ CoupleAndT ┆ RowNumb │
│ ar         ┆ de      ┆ ---      ┆ me       ┆     ┆ neDependen ┆ woDependen ┆ hreeOrMore ┆ er      │
│ ---        ┆ ---     ┆ i64      ┆ ---      ┆     ┆ t          ┆ ts         ┆ Dependents ┆ ---     │
│ i64        ┆ str     ┆          ┆ str      ┆     ┆ ---        ┆ ---        ┆ ---        ┆ i64     │
│            ┆         ┆          ┆          ┆     ┆ f64        ┆ f64        ┆ f64        ┆         │
╞════════════╪═════════╪══════════╪══════════╪═════╪════════════╪════════════╪════════════╪═════════╡
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ null       ┆ null       ┆ null       ┆ 14      │
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ 144.56     ┆ 144.56     ┆ 144.56     ┆ 14      │
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ 144.56     ┆ 144.56     ┆ 144.56     ┆ 15      │
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ null       ┆ null       ┆ null       ┆ 15      │
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ null       ┆ null       ┆ null       ┆ 16      │
└────────────┴─────────┴──────────┴──────────┴─────┴────────────┴────────────┴────────────┴─────────┘

In [ ]:
# selecting all columns
display(whats_my_footprint(df))

NameError: ignored

In [ ]:
%%time
display(df.schema)

{'BusinessYear': Int64,
 'StateCode': Utf8,
 'IssuerId': Int64,
 'SourceName': Utf8,
 'VersionNum': Int64,
 'ImportDate': Utf8,
 'IssuerId2': Int64,
 'FederalTIN': Utf8,
 'RateEffectiveDate': Utf8,
 'RateExpirationDate': Utf8,
 'PlanId': Utf8,
 'RatingAreaId': Utf8,
 'Tobacco': Utf8,
 'Age': Utf8,
 'IndividualRate': Float64,
 'IndividualTobaccoRate': Utf8,
 'Couple': Float64,
 'PrimarySubscriberAndOneDependent': Float64,
 'PrimarySubscriberAndTwoDependents': Float64,
 'PrimarySubscriberAndThreeOrMoreDependents': Float64,
 'CoupleAndOneDependent': Float64,
 'CoupleAndTwoDependents': Float64,
 'CoupleAndThreeOrMoreDependents': Float64,
 'RowNumber': Int64}

CPU times: user 4.5 ms, sys: 1.1 ms, total: 5.6 ms
Wall time: 10.8 ms


In [ ]:
whats_my_footprint(df)

8677.0 MB in the memory


In [ ]:
from pyarrow import compute as pc
from pyarrow import Table as pt

In [ ]:
def fix_datetime(x, column_name):
    replacement = pc.strptime(x.column(column_name), format='%Y-%m-%d %H:%M:%S', unit='s')
    position = x.column_names.index(column_name)
    x = x.set_column(position, column_name, replacement)
    return x

In [ ]:
def fix_categories(x, column_name):
    replacement = x[column_name].dictionary_encode()
    position = x.column_names.index[column_name]
    x = x.set_column(position, column_name, replacement)
    return x

In [ ]:
whats_my_footprint(df)

8677.0 MB in the memory


In [ ]:
df

shape: (12694445, 24)
┌────────────┬─────────┬──────────┬──────────┬─────┬────────────┬────────────┬────────────┬─────────┐
│ BusinessYe ┆ StateCo ┆ IssuerId ┆ SourceNa ┆ ... ┆ CoupleAndO ┆ CoupleAndT ┆ CoupleAndT ┆ RowNumb │
│ ar         ┆ de      ┆ ---      ┆ me       ┆     ┆ neDependen ┆ woDependen ┆ hreeOrMore ┆ er      │
│ ---        ┆ ---     ┆ i64      ┆ ---      ┆     ┆ t          ┆ ts         ┆ Dependents ┆ ---     │
│ i64        ┆ str     ┆          ┆ str      ┆     ┆ ---        ┆ ---        ┆ ---        ┆ i64     │
│            ┆         ┆          ┆          ┆     ┆ f64        ┆ f64        ┆ f64        ┆         │
╞════════════╪═════════╪══════════╪══════════╪═════╪════════════╪════════════╪════════════╪═════════╡
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ null       ┆ null       ┆ null       ┆ 14      │
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ 144.56     ┆ 144.56     ┆ 144.56     ┆ 14      │
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ 144.56     ┆ 144.56     ┆ 144.56     ┆ 15      │
│ 2014       ┆ AK      ┆ 21989    ┆ HIOS     ┆ ... ┆ null       ┆ null       ┆ null       ┆ 15      │
│ ...        ┆ ...     ┆ ...      ┆ ...      ┆ ... ┆ ...        ┆ ...        ┆ ...        ┆ ...     │
│ 2016       ┆ WV      ┆ 96480    ┆ SERFF    ┆ ... ┆ null       ┆ null       ┆ null       ┆ 2034    │
│ 2016       ┆ WV      ┆ 96480    ┆ SERFF    ┆ ... ┆ null       ┆ null       ┆ null       ┆ 2035    │
│ 2016       ┆ WV      ┆ 96480    ┆ SERFF    ┆ ... ┆ null       ┆ null       ┆ null       ┆ 2036    │
│ 2016       ┆ WV      ┆ 96480    ┆ SERFF    ┆ ... ┆ null       ┆ null       ┆ null       ┆ 2037    │
└────────────┴─────────┴──────────┴──────────┴─────┴────────────┴────────────┴────────────┴─────────┘

In [ ]:
%%time
df.groupby(["StateCode"], maintain_order=True).agg([pl.col("*").count()])

CPU times: user 565 ms, sys: 209 ms, total: 774 ms
Wall time: 477 ms


shape: (39, 24)
┌─────────┬────────────┬──────────┬──────────┬─────┬────────────┬────────────┬────────────┬─────────┐
│ StateCo ┆ BusinessYe ┆ IssuerId ┆ SourceNa ┆ ... ┆ CoupleAndO ┆ CoupleAndT ┆ CoupleAndT ┆ RowNumb │
│ de      ┆ ar         ┆ ---      ┆ me       ┆     ┆ neDependen ┆ woDependen ┆ hreeOrMore ┆ er      │
│ ---     ┆ ---        ┆ u32      ┆ ---      ┆     ┆ t          ┆ ts         ┆ Dependents ┆ ---     │
│ str     ┆ u32        ┆          ┆ u32      ┆     ┆ ---        ┆ ---        ┆ ---        ┆ u32     │
│         ┆            ┆          ┆          ┆     ┆ u32        ┆ u32        ┆ u32        ┆         │
╞═════════╪════════════╪══════════╪══════════╪═════╪════════════╪════════════╪════════════╪═════════╡
│ AK      ┆ 84412      ┆ 84412    ┆ 84412    ┆ ... ┆ 84412      ┆ 84412      ┆ 84412      ┆ 84412   │
│ AL      ┆ 196896     ┆ 196896   ┆ 196896   ┆ ... ┆ 196896     ┆ 196896     ┆ 196896     ┆ 196896  │
│ AZ      ┆ 350032     ┆ 350032   ┆ 350032   ┆ ... ┆ 350032     ┆ 350032     ┆ 350032     ┆ 350032  │
│ FL      ┆ 1702472    ┆ 1702472  ┆ 1702472  ┆ ... ┆ 1702472    ┆ 1702472    ┆ 1702472    ┆ 1702472 │
│ ...     ┆ ...        ┆ ...      ┆ ...      ┆ ... ┆ ...        ┆ ...        ┆ ...        ┆ ...     │
│ UT      ┆ 85398      ┆ 85398    ┆ 85398    ┆ ... ┆ 85398      ┆ 85398      ┆ 85398      ┆ 85398   │
│ NV      ┆ 32016      ┆ 32016    ┆ 32016    ┆ ... ┆ 32016      ┆ 32016      ┆ 32016      ┆ 32016   │
│ OR      ┆ 68402      ┆ 68402    ┆ 68402    ┆ ... ┆ 68402      ┆ 68402      ┆ 68402      ┆ 68402   │
│ HI      ┆ 1518       ┆ 1518     ┆ 1518     ┆ ... ┆ 1518       ┆ 1518       ┆ 1518       ┆ 1518    │
└─────────┴────────────┴──────────┴──────────┴─────┴────────────┴────────────┴────────────┴─────────┘

**Runtime for both polars and pandas to list out columns**

In [ ]:
%%time
df.columns

CPU times: user 21 µs, sys: 4 µs, total: 25 µs
Wall time: 28.4 µs


['BusinessYear',
 'StateCode',
 'IssuerId',
 'SourceName',
 'VersionNum',
 'ImportDate',
 'IssuerId2',
 'FederalTIN',
 'RateEffectiveDate',
 'RateExpirationDate',
 'PlanId',
 'RatingAreaId',
 'Tobacco',
 'Age',
 'IndividualRate',
 'IndividualTobaccoRate',
 'Couple',
 'PrimarySubscriberAndOneDependent',
 'PrimarySubscriberAndTwoDependents',
 'PrimarySubscriberAndThreeOrMoreDependents',
 'CoupleAndOneDependent',
 'CoupleAndTwoDependents',
 'CoupleAndThreeOrMoreDependents',
 'RowNumber']

In [ ]:
%%time
data.columns

CPU times: user 17 µs, sys: 0 ns, total: 17 µs
Wall time: 20.3 µs


Index(['BusinessYear', 'StateCode', 'IssuerId', 'SourceName', 'VersionNum',
       'ImportDate', 'IssuerId2', 'FederalTIN', 'RateEffectiveDate',
       'RateExpirationDate', 'PlanId', 'RatingAreaId', 'Tobacco', 'Age',
       'IndividualRate', 'IndividualTobaccoRate', 'Couple',
       'PrimarySubscriberAndOneDependent', 'PrimarySubscriberAndTwoDependents',
       'PrimarySubscriberAndThreeOrMoreDependents', 'CoupleAndOneDependent',
       'CoupleAndTwoDependents', 'CoupleAndThreeOrMoreDependents',
       'RowNumber'],
      dtype='object')

**Conclusion**

In overall, Polars manage to execute with less memory compared to Pandas. 